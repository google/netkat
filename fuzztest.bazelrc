### DO NOT EDIT. Generated file.
#
# To regenerate, run the following from your project's workspace:
#
#  bazel run @com_google_fuzztest//bazel:setup_configs > fuzztest.bazelrc
#
# And don't forget to add the following to your project's .bazelrc:
#
#  try-import %workspace%/fuzztest.bazelrc
### Common options.
#
# Do not use directly.

# Standard define for \"ifdef-ing\" any fuzz test specific code.
build:fuzztest-common --copt=-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION

# In fuzz tests, we want to catch assertion violations even in optimized builds.
build:fuzztest-common --copt=-UNDEBUG

# Enable libc++ assertions.
# See https://libcxx.llvm.org/UsingLibcxx.html#enabling-the-safe-libc-mode
build:fuzztest-common --copt=-D_LIBCPP_ENABLE_ASSERTIONS=1
### ASan (Address Sanitizer) build configuration.
#
# Use with: --config=asan

build:asan --linkopt=-fsanitize=address
build:asan --copt=-fsanitize=address
### FuzzTest build configuration.
#
# Use with: --config=fuzztest
#
# Note that this configuration includes the ASan configuration.

build:fuzztest --config=asan
build:fuzztest --config=fuzztest-common

# Link statically.
build:fuzztest --dynamic_mode=off

# We rely on the following flag instead of the compiler provided
# __has_feature(address_sanitizer) to know that we have an ASAN build even in
# the uninstrumented runtime.
build:fuzztest --copt=-DADDRESS_SANITIZER
