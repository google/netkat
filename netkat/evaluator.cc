// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

#include "netkat/evaluator.h"

#include "absl/container/flat_hash_set.h"
#include "absl/log/log.h"
#include "netkat/netkat.pb.h"

namespace netkat {

bool Evaluate(const PredicateProto& predicate, const Packet& packet) {
  switch (predicate.predicate_case()) {
    case PredicateProto::kBoolConstant:
      return predicate.bool_constant().value();
    case PredicateProto::kNotOp:
      return !Evaluate(predicate.not_op().negand(), packet);
    case PredicateProto::kAndOp:
      return Evaluate(predicate.and_op().left(), packet) &&
             Evaluate(predicate.and_op().right(), packet);
    case PredicateProto::kOrOp:
      return Evaluate(predicate.or_op().left(), packet) ||
             Evaluate(predicate.or_op().right(), packet);
    case PredicateProto::kXorOp:
      return (!Evaluate(predicate.xor_op().left(), packet) &&
              Evaluate(predicate.xor_op().right(), packet)) ||
             (Evaluate(predicate.xor_op().left(), packet) &&
              !Evaluate(predicate.xor_op().right(), packet));
    case PredicateProto::kMatch:
      if (auto iter = packet.find(predicate.match().field());
          iter != packet.end()) {
        return iter->second == predicate.match().value();
      } else {
        return false;
      }
    case PredicateProto::PREDICATE_NOT_SET:
      return false;
  }
  LOG(FATAL) << "Unexpected value for PredicateProto predicate_case: "
             << static_cast<int>(predicate.predicate_case());
}

absl::flat_hash_set<Packet> Evaluate(
    const PolicyProto& policy, const absl::flat_hash_set<Packet>& packets) {
  absl::flat_hash_set<Packet> result;
  for (const Packet& packet : packets) {
    result.merge(Evaluate(policy, packet));
  }
  return result;
}

absl::flat_hash_set<Packet> Evaluate(const PolicyProto& policy,
                                     const Packet& packet) {
  switch (policy.policy_case()) {
    case PolicyProto::kFilter:
      return Evaluate(policy.filter(), packet)
                 ? absl::flat_hash_set<Packet>({packet})
                 : absl::flat_hash_set<Packet>();
    case PolicyProto::kModification: {
      Packet modified_packet = packet;
      // Adds field if it doesn't exist, and modifies it otherwise.
      modified_packet[policy.modification().field()] =
          policy.modification().value();
      return {modified_packet};
    }
    case PolicyProto::kRecord:
      // Record is treated as a no-op.
      return {packet};
    case PolicyProto::kSequenceOp:
      return Evaluate(policy.sequence_op().right(),
                      Evaluate(policy.sequence_op().left(), packet));
    case PolicyProto::kUnionOp: {
      absl::flat_hash_set<Packet> result =
          Evaluate(policy.union_op().left(), packet);
      result.merge(Evaluate(policy.union_op().right(), packet));
      return result;
    }
    case PolicyProto::kIterateOp: {
      // p* = 1 + p + p;p + p;p;p + ...
      absl::flat_hash_set<Packet> result = {packet};  // 1
      // Evaluate p on result until fixed point, marked by no change in size.
      int last_size;
      do {
        last_size = result.size();
        result.merge(Evaluate(policy.iterate_op().iterable(), result));  // p^n
      } while (last_size != result.size());
      return result;
    }
    case PolicyProto::kPushOp: {
      return Evaluate(
          policy.push_op().policy(),
          Evaluate(policy.push_op().predicate(), packet) ? packet : Packet());
    }
    case PolicyProto::POLICY_NOT_SET:
      // Unset policy is treated as Deny.
      return {};
  }
}

}  // namespace netkat
