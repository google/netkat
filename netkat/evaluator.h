// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// -----------------------------------------------------------------------------
// File: evaluator.h
// -----------------------------------------------------------------------------
//
// Defines a library of functions for evaluating NetKAT predicates and policies
// on concrete packets.
//
// See go/netkat-hld for more details.

#ifndef GOOGLE_NETKAT_NETKAT_EVALUATOR_H_
#define GOOGLE_NETKAT_NETKAT_EVALUATOR_H_

#include <string>

#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "netkat/netkat.pb.h"

namespace netkat {

// A NetKAT packet is a map from field names to their values.
//
// NOTE: This is a simplistic, initial definition of `Packet` that we expect to
// replace in the future. In particular, the value type will change to support
// things like 128-bit IPv6 addresses.
//
// Fields that are not present in the map are assumed to carry an implicit "not
// present" value that is distinct from all explicitly-assignable values.
using Packet = absl::flat_hash_map<std::string, int>;

// Returns true if the given `packet` satisfies the given `predicate`, false
// otherwise.
//
// Note: Uninitialized predicates are considered unsatisfiable.
bool Evaluate(const PredicateProto& predicate, const Packet& packet);

// Returns the output packets produced by running the given policy on the given
// input packet. Treats `Record` (aka `dup`) as no-op and does not keep track of
// packet histories.
//
// Note: Uninitialized policies are considered DENY, returning the empty set.
absl::flat_hash_set<Packet> Evaluate(const PolicyProto& policy,
                                     const Packet& packet);

// Lifts policy evaluation to sets of packets.
absl::flat_hash_set<Packet> Evaluate(
    const PolicyProto& policy, const absl::flat_hash_set<Packet>& packets);

}  // namespace netkat

#endif  // GOOGLE_NETKAT_NETKAT_EVALUATOR_H_
