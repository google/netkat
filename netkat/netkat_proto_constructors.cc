// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "netkat/netkat_proto_constructors.h"

#include <string>
#include <utility>

#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "absl/strings/string_view.h"
#include "netkat/netkat.pb.h"

namespace netkat {

PredicateProto TrueProto() {
  PredicateProto proto;
  proto.mutable_bool_constant()->set_value(true);
  return proto;
}

PredicateProto FalseProto() {
  PredicateProto proto;
  proto.mutable_bool_constant()->set_value(false);
  return proto;
}

PredicateProto MatchProto(absl::string_view field, int value) {
  PredicateProto proto;
  PredicateProto::Match& match = *proto.mutable_match();
  match.set_field(std::string(field));
  match.set_value(value);
  return proto;
}
PredicateProto AndProto(PredicateProto left, PredicateProto right) {
  PredicateProto proto;
  PredicateProto::And& and_op = *proto.mutable_and_op();
  *and_op.mutable_left() = std::move(left);
  *and_op.mutable_right() = std::move(right);
  return proto;
}

PredicateProto OrProto(PredicateProto left, PredicateProto right) {
  PredicateProto proto;
  PredicateProto::Or& or_op = *proto.mutable_or_op();
  *or_op.mutable_left() = std::move(left);
  *or_op.mutable_right() = std::move(right);
  return proto;
}
PredicateProto NotProto(PredicateProto negand) {
  PredicateProto proto;
  PredicateProto::Not& not_op = *proto.mutable_not_op();
  *not_op.mutable_negand() = std::move(negand);
  return proto;
}
PredicateProto XorProto(PredicateProto left, PredicateProto right) {
  PredicateProto proto;
  PredicateProto::Xor& xor_op = *proto.mutable_xor_op();
  *xor_op.mutable_left() = std::move(left);
  *xor_op.mutable_right() = std::move(right);
  return proto;
}

// -- Basic Policy constructors ------------------------------------------------

PolicyProto FilterProto(PredicateProto filter) {
  PolicyProto policy;
  *policy.mutable_filter() = std::move(filter);
  return policy;
}

PolicyProto ModificationProto(absl::string_view field, int value) {
  PolicyProto policy;
  policy.mutable_modification()->set_field(field);
  policy.mutable_modification()->set_value(value);
  return policy;
}

PolicyProto RecordProto() {
  PolicyProto policy;
  policy.mutable_record();
  return policy;
}

PolicyProto SequenceProto(PolicyProto left, PolicyProto right) {
  PolicyProto policy;
  *policy.mutable_sequence_op()->mutable_left() = std::move(left);
  *policy.mutable_sequence_op()->mutable_right() = std::move(right);
  return policy;
}

PolicyProto UnionProto(PolicyProto left, PolicyProto right) {
  PolicyProto policy;
  *policy.mutable_union_op()->mutable_left() = std::move(left);
  *policy.mutable_union_op()->mutable_right() = std::move(right);
  return policy;
}

PolicyProto IterateProto(PolicyProto iterable) {
  PolicyProto policy;
  *policy.mutable_iterate_op()->mutable_iterable() = std::move(iterable);
  return policy;
}

PolicyProto DifferenceProto(PolicyProto left, PolicyProto right) {
  PolicyProto policy;
  *policy.mutable_difference_op()->mutable_left() = std::move(left);
  *policy.mutable_difference_op()->mutable_right() = std::move(right);
  return policy;
}

// -- Derived Policy constructors ----------------------------------------------

PolicyProto DenyProto() { return FilterProto(FalseProto()); }

PolicyProto AcceptProto() { return FilterProto(TrueProto()); }

std::string AsShorthandString(PredicateProto predicate) {
  switch (predicate.predicate_case()) {
    case PredicateProto::kBoolConstant:
      return predicate.bool_constant().value() ? "true" : "false";
    case PredicateProto::kMatch:
      return absl::StrFormat("@%s==%d", predicate.match().field(),
                             predicate.match().value());
    case PredicateProto::kAndOp:
      return absl::StrFormat("(%s && %s)",
                             AsShorthandString(predicate.and_op().left()),
                             AsShorthandString(predicate.and_op().right()));
    case PredicateProto::kOrOp:
      return absl::StrFormat("(%s || %s)",
                             AsShorthandString(predicate.or_op().left()),
                             AsShorthandString(predicate.or_op().right()));
    case PredicateProto::kNotOp:
      return absl::StrCat("!", AsShorthandString(predicate.not_op().negand()));
    case PredicateProto::kXorOp:
      return absl::StrFormat("(%s (+) %s)",
                             AsShorthandString(predicate.xor_op().left()),
                             AsShorthandString(predicate.xor_op().right()));
    case PredicateProto::PREDICATE_NOT_SET:
      return "false";
  }
}

std::string AsShorthandString(PolicyProto policy) {
  switch (policy.policy_case()) {
    case PolicyProto::kFilter:
      return AsShorthandString(policy.filter());
    case PolicyProto::kModification:
      return absl::StrFormat("@%s:=%d", policy.modification().field(),
                             policy.modification().value());
    case PolicyProto::kRecord:
      return "record";
    case PolicyProto::kSequenceOp:
      return absl::StrFormat("(%s; %s)",
                             AsShorthandString(policy.sequence_op().left()),
                             AsShorthandString(policy.sequence_op().right()));
    case PolicyProto::kUnionOp:
      return absl::StrFormat("(%s + %s)",
                             AsShorthandString(policy.union_op().left()),
                             AsShorthandString(policy.union_op().right()));
    case PolicyProto::kIterateOp:
      return absl::StrFormat("(%s)*",
                             AsShorthandString(policy.iterate_op().iterable()));
    case PolicyProto::kDifferenceOp:
      return absl::StrFormat("(%s - %s)",
                             AsShorthandString(policy.difference_op().left()),
                             AsShorthandString(policy.difference_op().right()));
    case PolicyProto::POLICY_NOT_SET:
      return "deny";
  }
}

}  // namespace netkat
