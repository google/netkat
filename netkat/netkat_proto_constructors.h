// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// -----------------------------------------------------------------------------
// File: netkat_proto_helpers.h
// -----------------------------------------------------------------------------
//
// Helper functions to make constructing netkat.proto messages more readable,
// specifically in unit test where readability is key.

#ifndef GOOGLE_NETKAT_NETKAT_NETKAT_PROTO_CONSTRUCTORS_H_
#define GOOGLE_NETKAT_NETKAT_NETKAT_PROTO_CONSTRUCTORS_H_

#include <string>

#include "absl/strings/string_view.h"
#include "netkat/netkat.pb.h"

namespace netkat {

// -- Predicate constructors ---------------------------------------------------

PredicateProto TrueProto();
PredicateProto FalseProto();
PredicateProto MatchProto(absl::string_view field, int value);
PredicateProto AndProto(PredicateProto left, PredicateProto right);
PredicateProto OrProto(PredicateProto left, PredicateProto right);
PredicateProto NotProto(PredicateProto negand);
PredicateProto XorProto(PredicateProto left, PredicateProto right);

// -- Basic Policy constructors ------------------------------------------------

PolicyProto FilterProto(PredicateProto filter);
PolicyProto ModificationProto(absl::string_view field, int value);
PolicyProto RecordProto();
PolicyProto SequenceProto(PolicyProto left, PolicyProto right);
PolicyProto UnionProto(PolicyProto left, PolicyProto right);
PolicyProto IterateProto(PolicyProto iterable);

// -- Derived Policy constructors ----------------------------------------------

PolicyProto DenyProto();
PolicyProto AcceptProto();

// Returns a shorthand string from a given NetKAT policy/predicate. This follows
// roughly the shorthand typically used in literature. Specifically:
//
//   Predicate And -> '&&'
//   Predicate Or -> '||'
//   Predicate Not -> '!'
//   Predicate Xor -> '(+)'
//   Policy Sequence -> ';'
//   Policy Or -> '+'
//   Iterate -> '*'
//   Record -> 'record'
//   Match -> '@field==value'
//   Modify -> '@field:=value'
//   True -> 'true'
//   False -> 'false'
//
// Note that parenthesis elimination is not performed.
//
// TODO(anthonyroy): Refactor this out to a different helper filer and/or widen
// the scope of this one.
std::string AsShorthandString(PolicyProto policy);
std::string AsShorthandString(PredicateProto predicate);

}  // namespace netkat

#endif  // GOOGLE_NETKAT_NETKAT_NETKAT_PROTO_CONSTRUCTORS_H_
