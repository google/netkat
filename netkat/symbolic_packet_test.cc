// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

#include "netkat/symbolic_packet.h"

#include "absl/base/no_destructor.h"
#include "absl/container/flat_hash_set.h"
#include "absl/strings/str_cat.h"
#include "fuzztest/fuzztest.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "gutil/status_matchers.h"
#include "netkat/evaluator.h"
#include "netkat/netkat_proto_constructors.h"

namespace netkat {
namespace {

using ::testing::StartsWith;

// We use a global manager object to exercise statefulness more deeply across
// test cases.
SymbolicPacketManager& Manager() {
  static absl::NoDestructor<SymbolicPacketManager> manager;
  return *manager;
}

// After executing all tests, we check once that no invariants are violated, for
// defense in depth. Checking invariants after each test (e.g. using a fixture)
// would likely not scale and seems overkill.
class CheckSymbolicPacketManagerInvariantsOnTearDown
    : public testing::Environment {
 public:
  ~CheckSymbolicPacketManagerInvariantsOnTearDown() override {}
  void SetUp() override {}
  void TearDown() override { ASSERT_OK(Manager().CheckInternalInvariants()); }
};
testing::Environment* const foo_env = testing::AddGlobalTestEnvironment(
    new CheckSymbolicPacketManagerInvariantsOnTearDown);

TEST(SymbolicPacketManagerTest, EmptySetIsEmptySet) {
  EXPECT_TRUE(Manager().IsEmptySet(Manager().EmptySet()));
  EXPECT_FALSE(Manager().IsFullSet(Manager().EmptySet()));
}

TEST(SymbolicPacketManagerTest, FullSetIsFullSet) {
  EXPECT_TRUE(Manager().IsFullSet(Manager().FullSet()));
  EXPECT_FALSE(Manager().IsEmptySet(Manager().FullSet()));
}

TEST(SymbolicPacketManagerTest, EmptySetDoesNotEqualFullSet) {
  EXPECT_NE(Manager().EmptySet(), Manager().FullSet());
}

TEST(SymbolicPacketManagerTest, AbslStringifyWorksForEmptySet) {
  EXPECT_THAT(absl::StrCat(Manager().EmptySet()), StartsWith("SymbolicPacket"));
}

TEST(SymbolicPacketManagerTest, AbslStringifyWorksForFullSet) {
  EXPECT_THAT(absl::StrCat(Manager().FullSet()), StartsWith("SymbolicPacket"));
}

TEST(SymbolicPacketManagerTest, AbslHashValueWorks) {
  absl::flat_hash_set<SymbolicPacket> set = {
      Manager().EmptySet(),
      Manager().FullSet(),
  };
  EXPECT_EQ(set.size(), 2);
}

TEST(SymbolicPacketManagerTest, TrueCompilesToFullSet) {
  EXPECT_EQ(Manager().Compile(TrueProto()), Manager().FullSet());
}

TEST(SymbolicPacketManagerTest, FalseCompilesToEmptySet) {
  EXPECT_EQ(Manager().Compile(FalseProto()), Manager().EmptySet());
}

void MatchCompilesToMatch(std::string field, int value) {
  EXPECT_EQ(Manager().Compile(MatchProto(field, value)),
            Manager().Match(field, value));
}
FUZZ_TEST(SymbolicPacketManagerTest, MatchCompilesToMatch);

void AndCompilesToAnd(const PredicateProto& left, const PredicateProto& right) {
  EXPECT_EQ(Manager().Compile(AndProto(left, right)),
            Manager().And(Manager().Compile(left), Manager().Compile(right)));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndCompilesToAnd);

void OrCompilesToOr(const PredicateProto& left, const PredicateProto& right) {
  EXPECT_EQ(Manager().Compile(OrProto(left, right)),
            Manager().Or(Manager().Compile(left), Manager().Compile(right)));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrCompilesToOr);

void NotCompilesToNot(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(NotProto(pred)),
            Manager().Not(Manager().Compile(pred)));
}
FUZZ_TEST(SymbolicPacketManagerTest, NotCompilesToNot);

void CompilationPreservesSemantics(const PredicateProto& pred,
                                   const Packet& packet) {
  EXPECT_EQ(Manager().Contains(Manager().Compile(pred), packet),
            Evaluate(pred, packet));
}
FUZZ_TEST(SymbolicPacketManagerTest, CompilationPreservesSemantics);

void EqualPredicatesCompileToEqualSymbolicPackets(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(pred), Manager().Compile(pred));
}
FUZZ_TEST(SymbolicPacketManagerTest,
          EqualPredicatesCompileToEqualSymbolicPackets);

void NegationCompilesToDifferentSymbolicPacket(const PredicateProto& pred) {
  EXPECT_NE(Manager().Compile(pred), Manager().Compile(NotProto(pred)));
}
FUZZ_TEST(SymbolicPacketManagerTest, NegationCompilesToDifferentSymbolicPacket);

void DoubleNegationCompilesToSameSymbolicPacket(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(pred),
            Manager().Compile(NotProto(NotProto(pred))));
}
FUZZ_TEST(SymbolicPacketManagerTest,
          DoubleNegationCompilesToSameSymbolicPacket);

TEST(SymbolicPacketManagerTest, TrueNotEqualsMatch) {
  EXPECT_NE(Manager().Compile(TrueProto()),
            Manager().Compile(MatchProto("hi", 42)));
}
TEST(SymbolicPacketManagerTest, FalseNotEqualsMatch) {
  EXPECT_NE(Manager().Compile(FalseProto()),
            Manager().Compile(MatchProto("hi", 42)));
}
TEST(SymbolicPacketManagerTest, MatchNotEqualsDifferentMatch) {
  EXPECT_NE(Manager().Compile(MatchProto("hi", 42)),
            Manager().Compile(MatchProto("bye", 42)));
  EXPECT_NE(Manager().Compile(MatchProto("hi", 42)),
            Manager().Compile(MatchProto("hi", 24)));
}
TEST(SymbolicPacketManagerTest, NotTrueEqualsFalse) {
  EXPECT_EQ(Manager().Compile(NotProto(TrueProto())),
            Manager().Compile(FalseProto()));
}

void AndIsIdempotent(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(AndProto(pred, pred)), Manager().Compile(pred));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndIsIdempotent);

void OrIsIdempotent(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(OrProto(pred, pred)), Manager().Compile(pred));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrIsIdempotent);

void PredOrItsNegationIsTrue(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(OrProto(pred, NotProto(pred))),
            Manager().Compile(TrueProto()));
}
FUZZ_TEST(SymbolicPacketManagerTest, PredOrItsNegationIsTrue);

void PredAndItsNegationIsFalse(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(AndProto(pred, NotProto(pred))),
            Manager().Compile(FalseProto()));
}
FUZZ_TEST(SymbolicPacketManagerTest, PredAndItsNegationIsFalse);

void AndTrueIsIdentity(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(AndProto(pred, TrueProto())),
            Manager().Compile(pred));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndTrueIsIdentity);

void OrFalseIsIdentity(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(OrProto(pred, FalseProto())),
            Manager().Compile(pred));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrFalseIsIdentity);

void AndFalseIsFalse(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(AndProto(pred, FalseProto())),
            Manager().Compile(FalseProto()));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndFalseIsFalse);

void OrTrueIsTrue(const PredicateProto& pred) {
  EXPECT_EQ(Manager().Compile(OrProto(pred, TrueProto())),
            Manager().Compile(TrueProto()));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrTrueIsTrue);

void AndIsCommutative(const PredicateProto& a, const PredicateProto& b) {
  EXPECT_EQ(Manager().Compile(AndProto(a, b)),
            Manager().Compile(AndProto(b, a)));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndIsCommutative);

void OrIsCommutative(const PredicateProto& a, const PredicateProto& b) {
  EXPECT_EQ(Manager().Compile(OrProto(a, b)), Manager().Compile(OrProto(b, a)));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrIsCommutative);

void DistributiveLawsHolds(const PredicateProto& a, const PredicateProto& b,
                           const PredicateProto& c) {
  EXPECT_EQ(Manager().Compile(AndProto(a, OrProto(b, c))),
            Manager().Compile(OrProto(AndProto(a, b), AndProto(a, c))));
  EXPECT_EQ(Manager().Compile(OrProto(a, AndProto(b, c))),
            Manager().Compile(AndProto(OrProto(a, b), OrProto(a, c))));
}
FUZZ_TEST(SymbolicPacketManagerTest, DistributiveLawsHolds);

void DeMorgansLawsHolds(const PredicateProto& a, const PredicateProto& b) {
  EXPECT_EQ(Manager().Compile(NotProto(AndProto(a, b))),
            Manager().Compile(OrProto(NotProto(a), NotProto(b))));
  EXPECT_EQ(Manager().Compile(NotProto(OrProto(a, b))),
            Manager().Compile(AndProto(NotProto(a), NotProto(b))));
}
FUZZ_TEST(SymbolicPacketManagerTest, DeMorgansLawsHolds);

void AndIsAssociative(const PredicateProto& a, const PredicateProto& b,
                      const PredicateProto& c) {
  EXPECT_EQ(Manager().Compile(AndProto(a, AndProto(b, c))),
            Manager().Compile(AndProto(AndProto(a, b), c)));
}
FUZZ_TEST(SymbolicPacketManagerTest, AndIsAssociative);

void OrIsAssociative(const PredicateProto& a, const PredicateProto& b,
                     const PredicateProto& c) {
  EXPECT_EQ(Manager().Compile(OrProto(a, OrProto(b, c))),
            Manager().Compile(OrProto(OrProto(a, b), c)));
}
FUZZ_TEST(SymbolicPacketManagerTest, OrIsAssociative);

}  // namespace
}  // namespace netkat
