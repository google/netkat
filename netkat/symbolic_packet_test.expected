================================================================================
Test case: p := (a=3 && b=4) || (b!=5 && c=5). Example from Katch paper Fig 3.
================================================================================
-- STRING ----------------------------------------------------------------------
SymbolicPacket<14>:
  InternedField<0>:'a' == 3 -> SymbolicPacket<13>
  InternedField<0>:'a' == * -> SymbolicPacket<6>
SymbolicPacket<13>:
  InternedField<1>:'b' == 4 -> SymbolicPacket<full>
  InternedField<1>:'b' == 5 -> SymbolicPacket<empty>
  InternedField<1>:'b' == * -> SymbolicPacket<5>
SymbolicPacket<6>:
  InternedField<1>:'b' == 5 -> SymbolicPacket<empty>
  InternedField<1>:'b' == * -> SymbolicPacket<5>
SymbolicPacket<5>:
  InternedField<2>:'c' == 5 -> SymbolicPacket<full>
  InternedField<2>:'c' == * -> SymbolicPacket<empty>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  14 [label="a"]
  14 -> 13 [label="3"]
  14 -> 6 [style=dashed]
  13 [label="b"]
  13 -> 4294967294 [label="4"]
  13 -> 4294967295 [label="5"]
  13 -> 5 [style=dashed]
  6 [label="b"]
  6 -> 4294967295 [label="5"]
  6 -> 5 [style=dashed]
  5 [label="c"]
  5 -> 4294967294 [label="5"]
  5 -> 4294967295 [style=dashed]
}
================================================================================
Test case: q := (b=3 && c=4) || (a=5 && c!=5). Example from Katch paper Fig 3.
================================================================================
-- STRING ----------------------------------------------------------------------
SymbolicPacket<24>:
  InternedField<0>:'a' == 5 -> SymbolicPacket<9>
  InternedField<0>:'a' == * -> SymbolicPacket<17>
SymbolicPacket<9>:
  InternedField<2>:'c' == 5 -> SymbolicPacket<empty>
  InternedField<2>:'c' == * -> SymbolicPacket<full>
SymbolicPacket<17>:
  InternedField<1>:'b' == 3 -> SymbolicPacket<16>
  InternedField<1>:'b' == * -> SymbolicPacket<empty>
SymbolicPacket<16>:
  InternedField<2>:'c' == 4 -> SymbolicPacket<full>
  InternedField<2>:'c' == * -> SymbolicPacket<empty>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  24 [label="a"]
  24 -> 9 [label="5"]
  24 -> 17 [style=dashed]
  9 [label="c"]
  9 -> 4294967295 [label="5"]
  9 -> 4294967294 [style=dashed]
  17 [label="b"]
  17 -> 16 [label="3"]
  17 -> 4294967295 [style=dashed]
  16 [label="c"]
  16 -> 4294967294 [label="4"]
  16 -> 4294967295 [style=dashed]
}
================================================================================
Test case: p + q. Example from Katch paper Fig 3.
================================================================================
-- STRING ----------------------------------------------------------------------
SymbolicPacket<34>:
  InternedField<0>:'a' == 3 -> SymbolicPacket<32>
  InternedField<0>:'a' == 5 -> SymbolicPacket<33>
  InternedField<0>:'a' == * -> SymbolicPacket<31>
SymbolicPacket<32>:
  InternedField<1>:'b' == 3 -> SymbolicPacket<30>
  InternedField<1>:'b' == 4 -> SymbolicPacket<full>
  InternedField<1>:'b' == 5 -> SymbolicPacket<empty>
  InternedField<1>:'b' == * -> SymbolicPacket<5>
SymbolicPacket<33>:
  InternedField<1>:'b' == 5 -> SymbolicPacket<9>
  InternedField<1>:'b' == * -> SymbolicPacket<full>
SymbolicPacket<31>:
  InternedField<1>:'b' == 3 -> SymbolicPacket<30>
  InternedField<1>:'b' == 5 -> SymbolicPacket<empty>
  InternedField<1>:'b' == * -> SymbolicPacket<5>
SymbolicPacket<30>:
  InternedField<2>:'c' == 4 -> SymbolicPacket<full>
  InternedField<2>:'c' == 5 -> SymbolicPacket<full>
  InternedField<2>:'c' == * -> SymbolicPacket<empty>
SymbolicPacket<5>:
  InternedField<2>:'c' == 5 -> SymbolicPacket<full>
  InternedField<2>:'c' == * -> SymbolicPacket<empty>
SymbolicPacket<9>:
  InternedField<2>:'c' == 5 -> SymbolicPacket<empty>
  InternedField<2>:'c' == * -> SymbolicPacket<full>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  34 [label="a"]
  34 -> 32 [label="3"]
  34 -> 33 [label="5"]
  34 -> 31 [style=dashed]
  32 [label="b"]
  32 -> 30 [label="3"]
  32 -> 4294967294 [label="4"]
  32 -> 4294967295 [label="5"]
  32 -> 5 [style=dashed]
  33 [label="b"]
  33 -> 9 [label="5"]
  33 -> 4294967294 [style=dashed]
  31 [label="b"]
  31 -> 30 [label="3"]
  31 -> 4294967295 [label="5"]
  31 -> 5 [style=dashed]
  30 [label="c"]
  30 -> 4294967294 [label="4"]
  30 -> 4294967294 [label="5"]
  30 -> 4294967295 [style=dashed]
  5 [label="c"]
  5 -> 4294967294 [label="5"]
  5 -> 4294967295 [style=dashed]
  9 [label="c"]
  9 -> 4294967295 [label="5"]
  9 -> 4294967294 [style=dashed]
}
